### PR TITLE
Jitclass support for '_' and '__' attrs, properties and methods

### PR DESCRIPTION
**Pull request by sklam**
Based on #1842 

Adds support for attribute, properties and methods with single and double leading underscores.

**Update 1**

The main change is adding the mangling of attribute name so that our struct helper in cgutils works (it treats attr starting with '_' differently.

### CODE REVIEW

File: numba/jitclass/boxing.py [hunk 1]
```diff
@@ -86,13 +86,12 @@ def _specialize_box(typ):
...
-            getter = _generate_getter(field)
-            setter = _generate_setter(field)
-            dct[field] = property(getter, setter)
+        getter = _generate_getter(field)
+        setter = _generate_setter(field)
+        dct[field] = property(getter, setter)
     # Inject properties as class properties
     for field, impdct in typ.jitprops.items():
-        if not field.startswith('_'):
+        if not field.startswith('__'):
```
Comment (pitrou) [inline numba/jitclass/boxing.py]: Why is this still needed?

File: numba/jitclass/boxing.py [hunk 2]
```diff
@@ -102,9 +101,8 @@ def _specialize_box(typ):
             dct[field] = property(getter, setter)
     # Inject methods as class members
     for name, func in typ.methods.items():
-        if not name.startswith('_'):
-            getter = _generate_method(name, func)
-            dct[name] = getter
+        if not (name.startswith('__') and name.endswith('__')):
```
Comment (pitrou) [inline numba/jitclass/boxing.py]: Ditto here. Also, why is there an `endswith` test (there isn't in line 94)?

File: numba/jitclass/boxing.py (hunk 1 above)
Comment (sklam) [inline numba/jitclass/boxing.py]: Right, this is not needed

File: numba/jitclass/boxing.py (hunk 2 above)
Comment (sklam) [inline numba/jitclass/boxing.py]: Special methods are ignored in this PR.  In #1865, there will be https://github.com/sklam/numba/blob/8e57a380abb6aba4c3bb90d206e03b0074107bbd/numba/jitclass/boxing.py#L120

File: numba/jitclass/base.py [hunk 3]
```diff
@@ -104,6 +113,19 @@ def _validate_spec(spec):
...
                             % (v,))
 
+
+def _fix_up_private_attr(clsname, spec):
+    """
+    Apply the same changes to dunder names as CPython would.
+    """
+    out = OrderedDict()
+    for k, v in spec.items():
+        if k.startswith('__'):
```
Comment (pitrou) [inline numba/jitclass/base.py]: I think you want to guard for `k.endswith('__')`, as `__special__` methods are not mangled.

**Comment by pitrou** [comment]
Thanks for the fixes! I'm merging now.

**Pull request merged**